graylog2-server/src/main/java/org/graylog2/indexer/datanode/DatanodeMigrationLockServiceImpl.java

```diff
@@ -0,0 +1,127 @@
+import org.graylog2.cluster.lock.Lock;
+import org.graylog2.cluster.lock.LockService;
+import org.graylog2.indexer.IndexSet;
+import org.graylog2.plugin.Tools;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.Executors;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.stream.Collectors;
+
+public class DatanodeMigrationLockServiceImpl implements DatanodeMigrationLockService {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DatanodeMigrationLockServiceImpl.class);
+    public static final int LOCK_EXTEND_PERIOD_SECONDS = 10;
+    private static final String LOCK_RESOURCE_PREFIX = "remote-reindex-migration_";
+
+    private final LockService lockService;
+
+
+    private final Set<Lock> activeLocks = Collections.synchronizedSet(new HashSet<>());
+
+    @Inject
+    public DatanodeMigrationLockServiceImpl(LockService lockService) {
+        this.lockService = lockService;
+        startLocksExtendingThread(lockService);
+    }
+
+    private void startLocksExtendingThread(LockService lockService) {
+        ScheduledExecutorService executorService = Executors.newSingleThreadScheduledExecutor(new ThreadFactoryBuilder().setNameFormat("migration-locks-service-backend-%d").setDaemon(true).setUncaughtExceptionHandler(new Tools.LogUncaughtExceptionHandler(LOG)).build());
+        executorService.scheduleAtFixedRate(() -> {
+            final Set<Optional<Lock>> extendedLocks = activeLocks.stream().map(lockService::extendLock).collect(Collectors.toSet());
+            if (!extendedLocks.isEmpty()) {
+                LOG.info("Extended TTL of {} datanode migration locks", extendedLocks.size());
+            }
+        }, LOCK_EXTEND_PERIOD_SECONDS, LOCK_EXTEND_PERIOD_SECONDS, TimeUnit.SECONDS);
+    }
+
+    @Override
+    public Lock acquireLock(IndexSet indexSet, Class<?> caller, String context, DatanodeMigrationLockWaitConfig config) {
+        final String indexSetID = indexSet.getConfig().id();
+        final String resource = LOCK_RESOURCE_PREFIX + indexSetID;
+        return waitForLock(resource, caller, context, indexSet, config);
+    }
+
+    @Override
+    public void tryRun(IndexSet indexSet, Class<?> caller, Runnable runnable) {
+        final Optional<Lock> lock = tryLock(indexSet, caller);
+        // here we have to keep the lock refreshed every now and then
+        lock.ifPresentOrElse(l -> {
+            try {
+                runnable.run();
+            } finally {
+                release(l);
+            }
+        }, () -> LOG.info("Couldn't enquire a lock of index set {}({}) for {}, skipping execution", indexSet.getConfig().title(), indexSet.getConfig().id(), caller.getName()));
+    }
+
+    @Override
+    public void release(Lock lock) {
```

Comment:
Because of separate thread accessing the `activeLocks` collection, I'd synchronize this method as well.

Reply:
:+1: done